### PR TITLE
show interstitial web page to allow user to choose IDP when multiple IDPs are configured and authorize endpoint query param to choose IDP is not used

### PR DESCRIPTION
**Pull request by cfryanr**
Addresses #1665.

The new page is styled exactly the same as the existing LDAP/AD login page, with the same fonts, borders, buttons, etc.

Here is a screenshot of what the new page looks like. Note that the rocket ship emoji's are part of the configured IDP `displayName` from the FederationDomain's `spec`, so those would not normally appear on the buttons. They are included only to ensure that unicode characters in the `displayName` works in this UI too.

Screenshot

Here is a screenshot of what the same page looks like when the browser has Javascript disabled. It still works, its just a little uglier.

Screenshot

**Release note**:

```release-note
Yes, TBD.
```

### CODE REVIEW

File: internal/testutil/assertions.go [hunk 1]
```diff
@@ -99,6 +99,19 @@ func RequireSecurityHeadersWithLoginPageCSPs(t *testing.T, response *httptest.Re
...
+	// Loosely confirm that the unique CSPs needed for the login page were used.
+	cspHeader := response.Header().Get("Content-Security-Policy")
+	require.Contains(t, cspHeader, "style-src '")  // loose assertion
+	require.Contains(t, cspHeader, "script-src '") // loose assertion
+	require.Contains(t, cspHeader, "style-src '")  // loose assertion
+	require.Contains(t, cspHeader, "img-src data:")
+	require.NotContains(t, cspHeader, "connect-src *") // only needed by form_post page
+
+	// Also require all the usual security headers.
+	requireSecurityHeaders(t, response)
```
Comment (benjaminapetersen) [inline internal/testutil/assertions.go:112]: Just skimmed `requireSecurityHeaders`, noticed `Strict-Transport-Security` isn't included.  Something we should consider adding it?

**Comment by benjaminapetersen** [comment]
With JS & Without JS styles look good.  My aesthetics want to remove the underline on the link, but I think for accessibility purposes its prob best to leave it. 👍

File: internal/federationdomain/endpoints/auth/auth_handler.go [hunk 2]
```diff
@@ -153,6 +166,20 @@ func NewHandler(
...
+	idpFinder federationdomainproviders.FederationDomainIdentityProvidersFinderI,
+	idpNameQueryParamValue string,
+	requestedBrowserlessFlow bool,
+) bool {
+	clientDidNotRequestSpecificIDP := len(idpNameQueryParamValue) == 0
+	clientRequestedBrowserBasedFlow := !requestedBrowserlessFlow
+	inBackwardsCompatMode := idpFinder.HasDefaultIDP()
+	federationDomainSpecHasSomeValidIDPs := idpFinder.IDPCount() > 0
+
+	return clientDidNotRequestSpecificIDP && clientRequestedBrowserBasedFlow &&
```
Comment (benjaminapetersen) [inline internal/federationdomain/endpoints/auth/auth_handler.go:179]: Nice, this reads great, no comment needed! :)

File: internal/testutil/assertions.go (hunk 1 above)
Comment (cfryanr) [reply internal/testutil/assertions.go:112]: I don't think so. That is a domain-wide setting, and we are potentially just one app within a domain. We avoid the issue by requiring TLS on all incoming connections by default.

File: internal/federationdomain/endpoints/auth/auth_handler.go [hunk 3]
```diff
@@ -89,6 +89,19 @@
...
 		// The Pinniped CLI has been sending these params since v0.9.0.
 		idpNameQueryParamValue := r.Form.Get(oidcapi.AuthorizeUpstreamIDPNameParamName)
+
+		// Check if we are in a special case where we should inject an interstitial page to ask the user
+		// which IDP they would like to use.
+		if shouldShowIDPChooser(idpFinder, idpNameQueryParamValue, requestedBrowserlessFlow) {
+			// Redirect to the IDP chooser page with all the same query/form params. When the user chooses an IDP,
+			// it will redirect back to here with all the same params again, with the pinniped_idp_name param added.
+			http.Redirect(w, r,
+				fmt.Sprintf("%s%s?%s", downstreamIssuer, oidc.ChooseIDPEndpointPath, r.Form.Encode()),
```
[the remark this answers is not in the document]
Comment (cfryanr) [reply internal/federationdomain/endpoints/auth/auth_handler.go:99]: The client is not allowed to influence the protocol, hostname, or path in this redirect. They can only influence the query params.

File: internal/federationdomain/endpoints/chooseidp/chooseidphtml/chooseidphtml_test.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	var buf bytes.Buffer
+	require.NoError(t, Template().Execute(&buf, pageInputs))
+	require.Equal(t, expectedHTML, buf.String())
+}
+
+func TestContentSecurityPolicy(t *testing.T) {
+	require.Equal(t, testExpectedCSP, ContentSecurityPolicy())
+}
+
+func TestCSS(t *testing.T) {
```
Comment (joshuatcasey) [inline internal/federationdomain/endpoints/chooseidp/chooseidphtml/chooseidphtml_test.go:59]: Add a `TestJS` unit test

File: internal/federationdomain/endpoints/chooseidp/choose_idp_handler_test.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		{
+			name:      "no query params on the request",
+			method:    http.MethodGet,
+			reqTarget: "/some/path" + oidc.ChooseIDPEndpointPath,
+			idps: oidctestutil.NewUpstreamIDPListerBuilder().
+				WithOIDC(oidctestutil.NewTestUpstreamOIDCIdentityProviderBuilder().WithName("x-some-idp").Build()).
+				BuildFederationDomainIdentityProvidersListerFinder(),
+			wantStatus:      http.StatusBadRequest,
+			wantContentType: "text/plain; charset=utf-8",
+			wantBodyString:  "Bad Request: missing required query params\n",
```
Comment (joshuatcasey) [inline internal/federationdomain/endpoints/chooseidp/choose_idp_handler_test.go:108]: We should tell the client which query params are missing. Easiest way is just to list all required parameters.

File: internal/federationdomain/endpoints/chooseidp/chooseidphtml/chooseidphtml_test.go (hunk 4 above)
Comment (cfryanr) [reply internal/federationdomain/endpoints/chooseidp/chooseidphtml/chooseidphtml_test.go:59]: Added in most recent force push (`0501159`).

File: internal/federationdomain/endpoints/chooseidp/choose_idp_handler_test.go (hunk 5 above)
Comment (cfryanr) [reply internal/federationdomain/endpoints/chooseidp/choose_idp_handler_test.go]: Added in most recent force push (`0501159`).

**Review by joshuatcasey — APPROVED** [review]

**Pull request merged**